### PR TITLE
Initial defer/reinstate flows added

### PR DESCRIPTION
**Pull request by jesseyuen**
This PR includes:

- updated records.json with deferral data
- defer and reinstate flows added (with some language to be considered)
- updates to the record _actions bar_ to ensure the links remain pinned to the right

✨ Special shout outs to @edwardhorsford for his support to get it to this point. ✨

### CODE REVIEW

File: app/routes.js [hunk 1]
```diff
@@ -428,7 +430,7 @@ router.get('/record/:uuid/trn', (req, res) => {
   }
 })
 
-// Manually advance an application from pending to trn received
+// Manually advance an application from Pending QTS received to QTS awarded
```
Comment (edwardhorsford) [inline app/routes.js]:
```suggestion
// Manually advance an application from Pending QTS to QTS awarded
```

File: app/routes.js [hunk 2]
```diff
@@ -464,6 +466,90 @@ router.post('/record/:uuid/qts/qts-recommended', (req, res) => {
   }
 })
 
+// Copy defer data back to real record
+router.post('/record/:uuid/defer/defer', (req, res) => {
```
Comment (edwardhorsford) [inline app/routes.js]: This url is a bit confusing. `Defer/confirm`? update?

File: app/routes.js [hunk 3]
```diff
@@ -464,6 +466,90 @@ router.post('/record/:uuid/qts/qts-recommended', (req, res) => {
...
+  const data = req.session.data
+  const newRecord = data.record
+
+  // Update failed or no data
+  if (!newRecord){
+    res.redirect('/record/:uuid')
+  }
+  else {
+    let radioChoice = newRecord.deferredDateRadio
+    if (radioChoice != "On another day") {
```
Comment (edwardhorsford) [inline app/routes.js]: We can actually get rid of this wrapping if statement - since the other ones are already specific.

File: app/routes.js [hunk 4]
```diff
@@ -464,6 +466,90 @@ router.post('/record/:uuid/qts/qts-recommended', (req, res) => {
...
+      if (radioChoice == "Yesterday") {
+        newRecord.deferredDate = filters.toDateArray(moment().subtract(1, "days"))
+      } 
+    }
+    res.redirect('/record/' + req.params.uuid + '/defer/confirm')
+  }
+})
+
+// Copy reinstate data back to real record
+router.post('/record/:uuid/reinstate/reinstate', (req, res) => {
```
Comment (edwardhorsford) [inline app/routes.js]: This url is confusing.

File: app/routes.js [hunk 5]
```diff
@@ -464,6 +466,90 @@ router.post('/record/:uuid/qts/qts-recommended', (req, res) => {
...
+})
+
+// Get dates for reinstate flow
+router.post('/record/:uuid/reinstate', (req, res) => {
+  const data = req.session.data
+  const newRecord = data.record
+
+  // Update failed or no data
+  if (!newRecord){
+    res.redirect('/record/:uuid')
```
Comment (edwardhorsford) [inline app/routes.js]:
```suggestion
    res.redirect(`/record/${req.params.uuid}`)
```

I spotted this is wrong in lots of places (my bad) - should be using `req.params.uuid`.

File: app/views/_includes/forms/defer.html [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+    legend: {
+      text: "When did the trainee defer?",
+      isPageHeading: true,
+      classes: "govuk-fieldset__legend--l govuk-!-margin-bottom-4"
+    }
+  },
+  items: [
+    {
+      text: "Today",
+      _value: todayDateArray
```
Comment (edwardhorsford) [inline app/views/_includes/forms/defer.html]: can delete this now

File: app/views/_includes/forms/defer.html [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+    },
+    {
+      text: "Yesterday",
+      _value: yesterdayDateArray
```
Comment (edwardhorsford) [inline app/views/_includes/forms/defer.html]: can delete this now

File: app/views/_includes/forms/defer.html [hunk 8]
```diff
@@ -0,0 +1,70 @@
...
+        {
+          name: "year",
+          classes: "govuk-input--width-4",
+          value: customDeferDateArray["2"]
+        }
+      ]
+  }) }}
+{% endset %}
+
+{% set todayDateArray = "" | today | toDateArray %}
```
Comment (edwardhorsford) [inline app/views/_includes/forms/defer.html]: can delete these

File: app/views/_includes/forms/reinstate.html [hunk 9]
```diff
@@ -0,0 +1,70 @@
...
+        {
+          name: "year",
+          classes: "govuk-input--width-4",
+          value: customReinstateDateArray["2"]
+        }
+      ]
+  }) }}
+{% endset %}
+
+{% set todayDateArray = "" | today | toDateArray %}
```
Comment (edwardhorsford) [inline app/views/_includes/forms/reinstate.html]: can delete these

File: app/views/_includes/forms/reinstate.html [hunk 10]
```diff
@@ -0,0 +1,70 @@
...
+    }
+  },
+  items: [
+    {
+      text: "Today",
+      _value: todayDateArray
+    },
+    {
+      text: "Yesterday",
+      _value: yesterdayDateArray
```
Comment (edwardhorsford) [inline app/views/_includes/forms/reinstate.html]: can delete these

File: app/views/_includes/summary-cards/defer-details.html [hunk 11]
```diff
@@ -0,0 +1,54 @@
...
+  ] %}
+
+{% set deferDetailsHtml %}
+  {{ govukSummaryList({
+    rows: deferDetailsRows
+  }) }}
+  
+{% endset %}
+
+{% set complete = true if (record.contactDetails | sectionIsCompleted ) %}
```
Comment (edwardhorsford) [inline app/views/_includes/summary-cards/defer-details.html]: I think you can delete rows 28-45 and the final endif. they're only relevant to 'add' flows.

File: app/views/_includes/summary-cards/reinstate-details.html [hunk 12]
```diff
@@ -0,0 +1,54 @@
...
+  ] %}
+
+{% set reinstateDetailsHtml %}
+  {{ govukSummaryList({
+    rows: reinstateDetailsRows
+  }) }}
+  
+{% endset %}
+
+{% set complete = true if (record.contactDetails | sectionIsCompleted ) %}
```
Comment (edwardhorsford) [inline app/views/_includes/summary-cards/reinstate-details.html]: as above, delete rows 28-45 and final endif

File: app/views/record/defer.html [hunk 13]
```diff
@@ -0,0 +1,14 @@
+{% extends "_templates/_record-form.html" %}
+
+{% set pageHeadingTraineeName = data.record.personalDetails | getShortName %}
+{# {% set formAction = "./defer/confirm" %} #}
+  
+{% block formContent %}
+
+  <h1 class="govuk-heading-l govuk-!-margin-bottom-0">
+    <span class="govuk-caption-xl">{{pageHeadingTraineeName}}</span>
```
Comment (edwardhorsford) [inline app/views/record/defer.html]:
```suggestion
    <span class="govuk-caption-l">{{pageHeadingTraineeName}}</span>
```

File: app/views/record/defer/confirm.html [hunk 14]
```diff
@@ -0,0 +1,32 @@
+{% extends "_templates/_record-form.html" %}
+{% set pageHeadingTraineeName = data.record.personalDetails | getShortName %}
+{% set pageHeading = "Check deferral details" %}
+
+{% set formAction = "./defer" %}
+
+{% block formContent %}
+  <h1 class="govuk-heading-l">
+    <span class="govuk-caption-xl">{{pageHeadingTraineeName}}</span>
```
Comment (edwardhorsford) [inline app/views/record/defer/confirm.html]:
```suggestion
    <span class="govuk-caption-l">{{pageHeadingTraineeName}}</span>
```

File: app/views/record/reinstate.html [hunk 15]
```diff
@@ -0,0 +1,14 @@
+{% extends "_templates/_record-form.html" %}
+
+{% set pageHeadingTraineeName = data.record.personalDetails | getShortName %}
+{# {% set formAction = "./reinstate/confirm" %} #}
+  
+{% block formContent %}
+
+  <h1 class="govuk-heading-l govuk-!-margin-bottom-0">
+    <span class="govuk-caption-xl">{{pageHeadingTraineeName}}</span>
```
Comment (edwardhorsford) [inline app/views/record/reinstate.html]:
```suggestion
    <span class="govuk-caption-l">{{pageHeadingTraineeName}}</span>
```

File: app/views/record/reinstate/confirm.html [hunk 16]
```diff
@@ -0,0 +1,32 @@
+{% extends "_templates/_record-form.html" %}
+{% set pageHeadingTraineeName = data.record.personalDetails | getShortName %}
+{% set pageHeading = "Check reinstatement details" %}
+
+{% set formAction = "./reinstate" %}
+
+{% block formContent %}
+  <h1 class="govuk-heading-l">
+    <span class="govuk-caption-xl">{{pageHeadingTraineeName}}</span>
```
Comment (edwardhorsford) [inline app/views/record/reinstate/confirm.html]:
```suggestion
    <span class="govuk-caption-l">{{pageHeadingTraineeName}}</span>
```

File: scripts/generate-records.js [hunk 17]
```diff
@@ -50,6 +50,19 @@ const generateFakeApplication = (params = {}) => {
...
   }
+  // If the status is Deferred 
+  if (status === 'Deferred') {
+    // Make sure deferral date is after submitted date
+    deferredDate = params.deferredDate || faker.date.between(
+      moment(),
+      moment(params.submittedDate))
+    
+    // Make sure deferral date is between submitted and updated date
+    deferredDate = params.deferredDate || faker.date.between(
```
Comment (edwardhorsford) [inline scripts/generate-records.js:56]: you seem to be setting `deferredDate` twice in this if block.

**Review by edwardhorsford — CHANGES_REQUESTED** [review]

File: scripts/generate-records.js (hunk 17 above)
Comment (jesseyuen) [reply scripts/generate-records.js:56]: Good shout. That first `deferredDate` is redundant.

**Review by edwardhorsford — APPROVED** [review]

**Pull request merged**